Refactor/bootloader cleanup

mvo5 opened the pull request:
cmd/snappy,snappy,partition,provisioning: refactor bootloader code to make it easier to work with

As discussed in the bugfix/no-more-ab-partition pull-request this is my attempt to simplify and cleanup the bootloader/partition code. Feedback and suggestions for further cleanup are very welcome.

chipaca commented:
LGTM!

niemeyer commented on partition/public.go:
```diff
@@ -19,17 +19,13 @@
+		return g.Dir()
 	}
 
 	return ""
```
This should probably return an error, as a hint to the call site that it can actually fail. It also feels like a slightly curious construct.. what use is it to have a bootloader dir without knowing that the bootloader is?  And why does it make sense to have it public, but not the rest of the API?  Honest questions. The only use case I've seen of BootloaderDir is in a piece of code saying it's terrible. :-)  Can we get rid of it?

niemeyer replied:
I've looked further, and found a single use of SetBootVar outside of `partition/`, within `snappy/kernel.go`. Why don't we just expose the bootloader interface as `Bootloader`, define a `FindBootloader` function returning `(Bootloader, error)`, and get rid of all of those wrappers (SetBootVar, GetBootVar, etc)?

niemeyer commented on partition/uboot.go:
```diff
@@ -28,51 +28,33 @@ import (
+}
+
+func (u *uboot) Dir() string {
+	return filepath.Join(dirs.GlobalRootDir, "/boot/uboot")
```
That piece of code also suggested having a single directory, such as `/boot/loader`. What prevents us from having it sooner?

niemeyer commented on partition/uboot.go:
```diff
@@ -28,51 +28,33 @@ import (
+	return filepath.Join(u.Dir(), "uEnv.txt")
+}
+
+func (u *uboot) fwEnvFile() string {
```
I suggest expanding "fw", or taking it off. It might refer to "firewall", "firmware", "forward", etc.

niemeyer commented on partition/utils.go:
```diff
@@ -26,38 +26,22 @@ import (
 	}
 
-	bytes, err := exec.Command(args[0], args[1:]...).Output()
+	output, err := exec.Command(args[0], args[1:]...).CombinedOutput()
```
It's usually a bad idea to grab the stderr when the output is oriented towards machine-parsing, unless you're specifically looking to pick errors up from that as well. If what you're looking for is error-reporting with proper error from the command (a good idea!), I'd suggest instead picking up both by setting `Stdout` and `Stderr` to `bytes.Buffer` values, returning stdout alone on success, and stderr inside the error value on failures. I'd also keep the original function name for both (`runCommand{,Impl}`).

niemeyer commented on partition/bootloader.go:
```diff
@@ -43,27 +42,18 @@ var (
 
-type bootloaderName string
-
 type bootLoader interface {
```
Can we rename this to bootloaderIface or similar?  The subtle upper casing there is a bit cryptic and unfriendly to usual conventions.

elopio commented on partition/utils_test.go:
```diff
@@ -28,20 +28,15 @@ type UtilsTestSuite struct {
+	output, err := runCommandWithStdoutImpl("sh", "-c", "printf 'foo\nbar'")
 	c.Assert(err, IsNil)
 	c.Assert(output, DeepEquals, "foo\nbar")
+
```
spurious line. I'm just doing a quick scan here so I have nothing more useful to say :)

mvo5 commented:
Thanks for this excellent review @niemeyer ! I addressed all points and will investigate the bootloader directory mount point now. I think it needs changes to ubuntu-device-flash, I have a look at this now.

niemeyer commented on partition/bootloader.go:
```diff
@@ -43,27 +42,22 @@ var (
-
+// BootloaderIface provides an interface to interact with the system
+// bootloader
+type BootloaderIface interface {
```
This can be just `Bootloader` now that it is public, since the casing doesn't conflict with anything else.

niemeyer commented:
Just a trivial above, and then LGTM once you're happy.

We need to tweak the summary+description after what we agree in the mailing list.
